### ISSUE TITLE
KEYB.COM and log discrepancies

### ISSUE THREAD

**Issue by Torinde** [dosbox-staging/dosbox-staging#3233]
### Are you using the latest Dosbox-Staging Version?

- [X] I have checked releases and am using the latest release.

### Different version than latest?

dosbox-staging-windows-msys2-x86_64-v0.81.0-alpha-2226-g2dff7

### What Operating System are you using?

Windows 10

### Have you checked that no other similar issue already exists?

- [X] I have searched and not found similar issues.

### A clear and concise description of what the bug is.

Testing latest build for the issues I mentioned at https://github.com/dosbox-staging/dosbox-staging/issues/3128#issuecomment-1825396175

1. LOG always says codepage is 3021, even if changed by KEYB.COM to 872 (and it's actually 872/855 - Cyrillic letters appears on the place of box characters that were 3021)
![image](https://github.com/dosbox-staging/dosbox-staging/assets/97228894/72844127-58c3-4d54-81f6-c579954242f1)
![image](https://github.com/dosbox-staging/dosbox-staging/assets/97228894/047e0ddf-556e-4603-ae01-d150d94cccf3)

2. KEYB bg442 complains about "file bg442 not found" (see above screenshot) and LOG has a message "DOS: The keyboard layout's country code 'bg442' does not have a corresponding country", but `KEYB bg442 855` reports success (although LOG gives the same warning) - but with codepage 872 (while LOG as always says 3021... so **user requests 855, KEYB states it provides 872, LOG states 3021 is used**, actually  is used 855 or 872 (I haven't checked with Euro sign which of the two)
![image](https://github.com/dosbox-staging/dosbox-staging/assets/97228894/520c512c-9fb4-4abb-9972-7aa80120d22c)
![image](https://github.com/dosbox-staging/dosbox-staging/assets/97228894/7f6d565d-2339-4466-8b44-7fbcce789f57)

3. On changes of codepage (but not keyboard layout) with KEYB - LOG doesn't get any message (expected was "LAYOUT: Loaded codepage XXX for detected language XX"), but maybe that's because of the "always 3021" effect (thus from this wrong perspective there was no change from certain command, thus nothing was logged?)

4. bg241 layout seems to have flipped shift/non-shift for the number keys - but since I don't recognize it at all (what's the origin/purpose of this layout?) I can't claim for sure - maybe the author wanted it like that:
![image](https://github.com/dosbox-staging/dosbox-staging/assets/97228894/6266a7ba-fb7f-4ca8-b165-1b3c5e7454ed)

### Steps to reproduce the behaviour.

Explain how to reproduce
1. Start in portable mode with empty dosbox-staging.conf
2. enter keyb commands as per the screenshots

### Provide a Log

```text
2023-12-18 11:33:03.650 | DOS: Loaded codepage 3021 for detected language 'bg'
2023-12-18 11:33:03.672 | DOS: Loaded modern locale for country 359, 'Bulgaria'
2023-12-18 11:34:08.457 | DOS: Loaded codepage 3021 for detected language 'bg103'
2023-12-18 11:34:51.149 | DOS: Loaded codepage 3021 for detected language 'bg'
2023-12-18 11:35:05.959 | DOS: Loaded codepage 3021 for detected language 'bg241'
2023-12-18 11:56:02.961 | DOS: The keyboard layout's country code 'bg442' does not have a corresponding country
```

### Code of Conduct & Contributing Guidelines

- [X] Yes, I agree.

**Comment by FeralChild64** [comment]
@Torinde For the future: it is much easier, if a single issue addresses a single problem. 

Problems (1) and (3) are just missing/misleading logs, right?

Problem (2) `KEYB bg442` is definitely a separate issue - I can reproduce it also with 0.80.1. BTW, we have no bg442 keyboard layout (see https://github.com/dosbox-staging/dosbox-staging/blob/main/contrib/resources/freedos-keyboard/DOC/LAYOUTS.TXT). Still, the behavior seems strange to me.

Problem (4) `bg241` layout - unfortunately, I'm not familiar with such keyboard layouts. Would it be possible for you to install FreeDOS 1.3 under QEMU or VirtualBox and check the behavior there? if it behaves the same, than this issue should be reported upstream (we are using FreeDOS console fonts and keyboard definitions in DOSBox).

**Comment by FeralChild64** [comment]
@Torinde Please also look at issue https://github.com/dosbox-staging/dosbox-staging/issues/1925 - different code page loaded and different printed in some cases is a long standing issue, not limited to Bulgarian keyboards.

**Comment by johnnovak** [comment]
@Torinde I agree with the other guys, please don't overload issue tickets like this in the future. It's very hard to keep track of the different but related problems in such mega tickets and it's just overwhelming.

Open separate, more focused tickets next time, and link them via the "related issues" section please. Thank you.

**Comment by Torinde** [comment]
dosbox-staging-windows-msys2-x86_64-v0.81.0-alpha-2248-g91e65
![image](https://github.com/dosbox-staging/dosbox-staging/assets/97228894/046b07fb-5c79-45bc-b102-5e522cc2e193)

Corrections needed:
- bg103 855/872 - currently load 3021 instead of what user requested
- bg241
  - to load the codepage specified by the user (currently it always loads 872)
  - to load 3021 when no codepage is specified
- bg442 
  - to work also when no codepage is specified (loading 3021)
  - to remove issue that makes LOG complain about corresponding country when codepage specified by user is not 872
  - to load the codepage specified by the user (currently it always loads 872)
- bg when using 3021 - `Shift+Q` proper output is `ы` (Russian letter) and not Greek delta `δ`.

Modernization/Improvements possible:
- default layout for Bulgaria (result from `keyb bg`) to be bg103 (currently it's bg442) - https://github.com/dosbox-staging/dosbox-staging/pull/4597#issuecomment-3539187918
- bg103 (or a new layout) - like on [Windows Phonetic Traditional](https://learn.microsoft.com/en-us/globalization/keyboards/kbdbgph1)
  - `Shift+3` to give `№` 
  - `Shift+6` to give `€`
  - `Shift+7` to give `§`
- bg442 (or a new layout) - like on [Windows Bulgarian](https://learn.microsoft.com/en-us/globalization/keyboards/kbdbulg)
  - `Shift+9` to give `–` (EN-DASH U+2013)
  - `Shift+-` to give `€`
  - `Shift+=` to give `$`

**Comment by rderooy** [comment]
Possibly useful:
https://archive.org/details/os-2-keyboards-and-codepages/page/6/mode/2up?view=theater
![image](https://github.com/dosbox-staging/dosbox-staging/assets/17701098/297ffaba-3e32-4219-87c8-39cf3a6ad6df)

https://archive.org/details/os-2-keyboards-and-codepages/page/8/mode/2up?view=theater
![image](https://github.com/dosbox-staging/dosbox-staging/assets/17701098/4dab2f43-1139-4891-a593-711ec699281e)

This one clearly has the numbers inverted...

**Comment by johnnovak** [comment]
@FeralChild64 Leave this ticket open or kill it with _extreme prejudice_? Not sure if it's a legit issue or just another Torinde special 😆

**Comment by FeralChild64** [comment]
I am just ignoring this ticket for now.

Until I rewrite the keyboard layout loading/handling (like I did with the CPI files) it makes no sense to dig into it - and I don’t want to waste literally hours on discussions, I’m rather short of spare time recently.

**Comment by johnnovak** [comment]
@FeralChild64 Can we close this?

**Comment by FeralChild64** [comment]
I still didn't rewrite the keyboard layout handling, this is not trivial and needs some investigation and quite some testing. But if you really want to reduce the number of open tickets, we can do so and make a note in https://github.com/dosbox-staging/dosbox-staging/issues/1925, this has most likely the same root cause.

**Comment by johnnovak** [comment]
> I still didn't rewrite the keyboard layout handling, this is not trivial and needs some investigation and quite some testing. But if you really want to reduce the number of open tickets, we can do so and make a note in [#1925](https://github.com/dosbox-staging/dosbox-staging/issues/1925), this has most likely the same root cause.

Thanks, I just periodically go through all tickets to prune the backlog where I can.

You can do either; close this and make a note in #1925 , or mention this ticket there so we'll remember to close it.

Cheers!
